File: Scala/algorithmic_toolbox/src/main/scala-2.12/dynamicprogramming/EditDistance.scala
package dynamicprogramming

object EditDistance {
  def calcLevenshteinDistance(string1: String, string2: String): Int = {
    val n: Int = string1.length
    val m: Int = string2.length
    val editDistance: Array[Array[Int]] = Array.fill(n + 1, m + 1)(0)
    def matchingScore(c1: Char, c2: Char): Int = if (c1 == c2) 0 else 1

    for { ix <- 1 to n } editDistance(ix)(0) = ix
    for { jy <- 1 to m } editDistance(0)(jy) = jy

    for {
      ix <- 1 to n
      jy <- 1 to m
    } {
      val deletion: Int = editDistance(ix - 1)(jy) + 1
      val insertion: Int = editDistance(ix)(jy - 1) + 1
      val matching: Int = editDistance(ix - 1)(jy - 1) + matchingScore(string1(ix - 1), string2(jy - 1))
      editDistance(ix)(jy) = List(insertion, deletion, matching).min
    }

    editDistance(n)(m)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(string1, string2): List[String] = reader.take(2).toList
    val result: Int = calcLevenshteinDistance(string1, string2)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/dynamicprogramming/edit_distance.ml
open Batteries

let calc_Levenshtein_distance string1 string2 =
    let n = String.length string1 in
    let m = String.length string2 in
    let edit_distance = Array.make_matrix (n + 1) (m + 1) 0 in
    let matching_score c1 c2 = if c1 = c2 then 0 else 1 in
    for ix = 1 to n do
        edit_distance.(ix).(0) <- ix
    done;
    for jy = 1 to m do
        edit_distance.(0).(jy) <- jy
    done;

    for ix = 1 to n do
        for jy = 1 to m do
            let deletion = edit_distance.(ix - 1).(jy) + 1 in
            let insertion = edit_distance.(ix).(jy - 1) + 1 in
            let matching = edit_distance.(ix  -1).(jy - 1) +
                matching_score (String.get string1 (ix - 1)) (String.get string2 (jy - 1)) in
            edit_distance.(ix).(jy) <- List.min [deletion; insertion; matching]
        done;
    done;
    edit_distance.(n).(m)


let () =
    let string1 = read_line() in
    let string2 = read_line() in
    let result = calc_Levenshtein_distance string1 string2 in
    print_int result;
    print_newline()
